packages/contracts/contracts/L1/AssetManager.sol

```diff
@@ -0,0 +1,1470 @@
+            }
+
+            unchecked {
+                vault.asset.totalKro -= arr[0].mulDiv(amountToSlashOrAdd, totalAmount);
+                vault.reward.boostedReward -= arr[1].mulDiv(amountToSlashOrAdd, totalAmount);
+                vault.asset.totalPendingAssets -= arr[2].mulDiv(amountToSlashOrAdd, totalAmount);
+                vault.asset.totalPendingBoostedRewards -= arr[3].mulDiv(
+                    amountToSlashOrAdd,
+                    totalAmount
+                );
+                vault.reward.validatorRewardKro -= arr[4].mulDiv(amountToSlashOrAdd, totalAmount);
+            }
+
+            uint128 tax = amountToSlashOrAdd.mulDiv(TAX_NUMERATOR, TAX_DENOMINATOR);
+            unchecked {
+                amountToSlashOrAdd -= tax;
+            }
+            bool success = IERC20(GOVERNANCE_TOKEN).transfer(SECURITY_COUNCIL, tax);
+            require(success, "AssetManager: Transfer failed");
+
+            return amountToSlashOrAdd;
+        } else {
+            unchecked {
+                vault.asset.totalKro += arr[0].mulDiv(amountToSlashOrAdd, totalAmount);
+                vault.reward.boostedReward += arr[1].mulDiv(amountToSlashOrAdd, totalAmount);
+                vault.asset.totalPendingAssets += arr[2].mulDiv(amountToSlashOrAdd, totalAmount);
+                vault.asset.totalPendingBoostedRewards += arr[3].mulDiv(
+                    amountToSlashOrAdd,
+                    totalAmount
+                );
+                vault.reward.validatorRewardKro += arr[4].mulDiv(amountToSlashOrAdd, totalAmount);
+            }
+
+            return amountToSlashOrAdd;
+        }
+    }
+
+    /**
+     * @notice Internal function to get the boosted reward with the number of KGH.
+     *
+     * @param numKgh The number of KGH.
+     *
+     * @return The boosted reward with the number of KGH.
+     */
+    function _getBoostedReward(uint128 numKgh) internal view virtual returns (uint128) {
+        uint128 coefficient = BASE_REWARD.mulDiv(BOOSTED_REWARD_NUMERATOR, BOOSTED_REWARD_DENOM);
+        return uint128(Atan2.atan2(numKgh, 100).mulDiv(coefficient, 1 << 40));
+    }
+
+    /**
+     * @notice Internal function to try to remove the validator from the weight tree.
+     * 
+     * @param validator Address of the validator.
+     */
+    function _tryRemoveFromWeightTree(address validator) internal {
+        if (_vault[validator].asset.validatorKro < MIN_START_AMOUNT) {
+            _weightTree.remove(validator);
+        }
+    }
+
+    /**
+     * @notice Internal function to update the weight tree of the validator.
+     *
+     * @param validator Address of the validator.
+     */
+    function _updateWeightTree(address validator) internal {
+        _weightTree.update(
+            validator,
+            uint32(
+                (_vault[validator].asset.totalKro +
+                    _vault[validator].reward.boostedReward +
+                    _vault[validator].reward.validatorRewardKro) / 1e16
```

Comment:
Are there any reason to divide with `1e16`?

Reply:
As my first thought, dividing in`1e17` may result in a problem that it cannot reflect the relatively small amount of delegations, and the case you mentioned is not very likely to happen.

But I think it would be better to lower this to `1e13` and increase the bits of parameters in `Node` struct at `BalancedWeightTree` like this:
`struct Node {
        address addr;
        uint40 weight;
        uint40 weightSum;
        uint40 parent;
        uint40 leftChild;
        uint32 rightChild;
        bool isLeftChild;
    }
`
It still occupies 2 slots, so I think the gas will not increase much. How do you think about this? If reasonable, I can update the weight tree with the fix.

Reply:
That's a good idea. Then how about like below and just remove the denominator?

```solidity
    struct Node {
        address addr;
        uint32 parent;
        uint32 leftChild;
        uint32 rightChild;
        bool isLeftChild;
        uint120 weight;
        uint120 weightSum;
    }
```

Reply:
That looks much better, I'll fix it.